utils/ixdtf/src/parsers/mod.rs

```diff
@@ -0,0 +1,327 @@
+    ///
+    /// [temporal-md]: https://tc39.es/proposal-temporal/#prod-TemporalMonthDayString
+    pub fn parse_month_day(&mut self) -> ParserResult<IsoParseRecord> {
+        let md = datetime::parse_month_day(&mut self.cursor);
+
+        let Ok(month_day) = md else {
+            self.cursor.pos = 0;
+            return datetime::parse_annotated_date_time(DateTimeFlags::empty(), &mut self.cursor);
+        };
+
+        let calendar = if self.cursor.check_or(false, is_annotation_open) {
+            let set = annotations::parse_annotation_set(false, &mut self.cursor)?;
+            set.calendar
+        } else {
+            None
+        };
+
+        self.cursor.close()?;
+
+        Ok(IsoParseRecord {
+            date: DateRecord {
+                year: 0,
+                month: month_day.0,
+                day: month_day.1,
+            },
+            time: None,
+            tz: None,
+            calendar,
+        })
+    }
+
+    /// Parses the source as a [Time string][temporal-time].
+    ///
+    /// This method will parse as an annotated time string or annotated date time string.
+    ///
+    /// [temporal-time]: https://tc39.es/proposal-temporal/#prod-TemporalTimeString
+    pub fn parse_time(&mut self) -> ParserResult<IsoParseRecord> {
+        match time::parse_annotated_time_record(&mut self.cursor) {
+            Ok(None) => {
+                self.cursor.pos = 0;
+                datetime::parse_annotated_date_time(DateTimeFlags::TIME_REQ, &mut self.cursor)
+            }
+            Ok(Some(result)) => Ok(result),
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Parses the source as a Time Zone string.
+    pub fn parse_time_zone(&mut self) -> ParserResult<TimeZone> {
+        time_zone::parse_time_zone(&mut self.cursor)
+    }
+}
+
+/// A parser for Duration strings.
+///
+/// # Exmaple
+///
+/// ```
+/// use ixdtf::parsers::{IsoDurationParser, records::DurationParseRecord };
+///
+/// let duration_str = "P1Y2M3W1D";
+///
+/// let result = IsoDurationParser::new(duration_str).parse().unwrap();
+///
+/// assert_eq!(result.years, 1);
+/// assert_eq!(result.months, 2);
+/// assert_eq!(result.weeks, 3);
+/// assert_eq!(result.days, 1);
+///
+/// ```
+#[derive(Debug)]
+pub struct IsoDurationParser {
+    cursor: Cursor,
+}
+
+impl IsoDurationParser {
+    /// Creates a new `IsoDurationParser` from a target `&str`.
+    pub fn new(target: &str) -> Self {
+        Self {
+            cursor: Cursor::new(target),
+        }
+    }
+
+    /// Parse the contents of this `IsoDurationParser` into a `DurationParseRecord`.
+    pub fn parse(&mut self) -> ParserResult<DurationParseRecord> {
+        parse_duration(&mut self.cursor)
+    }
+}
+
+// ==== Mini cursor implementation for Iso8601 targets ====
+
+/// `Cursor` is a small cursor implementation for parsing Iso8601 grammar.
+#[derive(Debug)]
+pub(crate) struct Cursor {
+    pos: u32,
+    source: Vec<char>,
+}
```

Comment:
Suggestion: Ideally `Cursor` would borrow the `str`, like

```rust
pub(crate) struct Cursor<'a> {
    pos: u32,
    source: &'a str,
}
```

This would allow the `slice` function to return `&'a str`, like

```rust
impl<'a> Cursor {
    fn slice(&self, start: usize, end: usize) -> Option<&'a str> {
        self.source.get(start..end)
    }
}
```

This is preferable because we reduce memory allocations, which are slow especially in low-level parser code.

The only reason to convert to a `Vec<char>` would be if you need random access. However, I don't _think_ you need this, because

1. I think you are almost always stepping forwards one code point at a time. You can do this in UTF-8 without needing true random access.
2. IXDTF strings should be ASCII, right? So if you ever encounter a non-ASCII code point, you can error out, and then if you do need random access, you can use byte indices anyway.

Reply:
Yeah, agreed. This is the one improvement that I really think should happen as well. I'll take a look at it.

The one point that is going to be a bit weird as far as Temporal compatibility is the [`TemporalSign`](https://tc39.es/ecma262/#prod-TemporalSign), which isn't completely ASCII.

```txt
[TemporalSign] :::
    ASCIISign
    <MINUS>
```

Reply:
Acknowledged that there might be a non-ASCII minus sign. So that invalidates point 2, but I think point 1 should still apply. You are pretty much only ever walking forwards, so you shouldn't ever need to do random access. Taking slices on byte indices should be fine.